Refactor Rag notebook

richardsliu opened the pull request:
Refactored the RAG notebook to be more modular and documented. 

* Moved the CloudSQL code into the notebook instead of running on a Ray Clusters
* Execute remote Ray tasks instead of using a job submit with busy waiting
* Modified the CloudSQL code to use bulk insert

Fixes https://github.com/GoogleCloudPlatform/ai-on-gke/issues/425

Preview: https://github.com/GoogleCloudPlatform/ai-on-gke/blob/rag-notebook/applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb

andrewsykim commented on modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:
```diff
@@ -88,6 +88,7 @@ singleuser:
   extraEnv:
     # Used for GCSFuse to set the ephemeral storage as the home directory. If not set, it will show a permission error on the pod log when using GCSFuse.
     JUPYTER_ALLOW_INSECURE_WRITES: "true"
+    CLOUDSQL_INSTANCE_CONNECTION_NAME: ${cloudsql_instance_connection_name}
```
btw. this is going to add the CLOUDSQL env var and secrets for the Jupyter-only installation (not RAG)

andrewsykim replied:
(likewise for the secret mounts)

andrewsykim replied:
^ we spoke offline -- I meant that jupyer only installs will set these environment varibles

richardsliu replied:
Made these variables default to empty string. If not explicitly passed, the env var will be empty.

andrewsykim commented on applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb:
```diff
@@ -0,0 +1,449 @@
+   "outputs": [],
+   "source": [
+    "!pip install ray[default]==2.9.3 kaggle==1.6.6\n",
+    "!pip install langchain==0.1.10 ray==2.9.3 datasets sentence-transformers\n",
```
I think there were some issues in the past with some of these packages taking way too long to install. Specifically sentence-transformers. Are we going to bake the dependencies into the jupyter image?

richardsliu replied:
I've tested and it turns out the notebook does need to install langchain and sentence-transformers. I can see if we could use a custom image here to skip these pip installs.

richardsliu replied:
Changed to a custom image with dependencies baked in. Removed this section from the notebook.

andrewsykim commented on applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb:
```diff
@@ -0,0 +1,449 @@
+{
```
Can you name this file `rag-kaggle-ray-sql-interactive.ipynb` instead?

andrewsykim reviewed (commented):
Overall LGTM, this is a great improvement!

I think the following issues need to be addressed if we're going to cherry-pick this into release-1.1:
* Dependency install time and whether a custom image is warranted. 
* Leaking CloudSQL env vars and secret mounts into all Jupyter installments -- these should only be set when using Jupyter for RAG.
* If we choose to keep both versions of the notebook, please add the markdown descriptions in the other notebook as well

richardsliu commented:
> * Dependency install time and whether a custom image is warranted.

The notebook needs to install `sentence-transformer` and `langchain`. I'll see if we can use a custom image to reduce the install time.

> * Leaking CloudSQL env vars and secret mounts into all Jupyter installments -- these should only be set when using Jupyter for RAG (likewise for Ray ref: [CLOUDSQL_INSTANCE_CONNECTION_NAME env in RayCluster should only be configured for RAG deployments   #435](https://github.com/GoogleCloudPlatform/ai-on-gke/issues/435)).

This is now fixed.

> * If we choose to keep both versions of the notebook, please add the markdown descriptions in the other notebook as well

I prefer to keep this PR limited to changes relevant for the new notebook. Adding markdown to the older notebook wouldn't be too interesting since all the code is executed in one cell.

richardsliu commented on applications/rag/example_notebooks/rag-kaggle-ray-sql-refactored.ipynb:
```diff
@@ -0,0 +1,449 @@
+    "ray.init(\n",
+    "    address=\"ray://ray-cluster-kuberay-head-svc:10001\",\n",
+    "    runtime_env={\n",
+    "        \"pip\": [               \n",
```
I think Ray will skip pip installing these if the image is already on the node. When I tested this, this step usually finishes in a few seconds.

imreddy13 commented on applications/rag/example_notebooks/rag-kaggle-ray-sql-interactive.ipynb:
```diff
@@ -0,0 +1,425 @@
+    "\n",
+    "## Setup Kaggle Credentials\n",
+    "\n",
+    "First we will setup your Kaggle credentials. Replace the following with your own settings from the Kaggle web page. Navigate to https://www.kaggle.com/settings/account and generate an API token to be used to setup the env variable. See https://www.kaggle.com/docs/api#authentication how to create one."
```
nit: "to be used to setup the env variable" is kind of weird. Maybe just have the instructions for how to create Kaggle creds and ask users to replace the specific variables (KAGGLE_USERNAME and KAGGLE_KEY) in the cell below with their Kaggle creds?

richardsliu replied:
Merged the cells.

imreddy13 commented on modules/jupyter/jupyter_config/config-selfauth-autopilot.yaml:
```diff
@@ -111,9 +112,17 @@ singleuser:
     extraVolumeMounts:
         - name: data-vol
-          mountPath: /persist-data
+          mountPath: /data
```
we don't need persist-data anymore? awesome! this always confused me.

richardsliu replied:
Glad you agree.

imreddy13 commented on modules/jupyter/main.tf:
```diff
@@ -17,6 +17,7 @@ data "google_project" "project" {
 }
 
 locals {
+  cloudsql_instance_connection_name = var.cloudsql_instance_name != "" ? format("%s:%s:%s", var.project_id, var.db_region, var.cloudsql_instance_name) : ""
```
could you please try to run tf apply from /applications/jupyter to ensure that the secrets mounting isn't breaking just the jupyter installation? it should be fairly quick to test

richardsliu replied:
Good call. Fixed an error.

imreddy13 commented on modules/jupyter/variables.tf:
```diff
@@ -17,6 +17,18 @@ variable "namespace" {
+variable "notebook_image" {
+  type        = string
+  description = "Jupyter notebook image name"
+  default     = "jupyter/tensorflow-notebook"
```
In addition to defaulting, can you also explicitly set this value from /applications/jupyter/main.tf so it's clear which app is using which image?

richardsliu replied:
Done.

imreddy13 commented on modules/jupyter/variables.tf:
```diff
@@ -17,6 +17,18 @@ variable "namespace" {
+variable "notebook_image_tag" {
+  type        = string
+  description = "Jupyter notebook image tag"
+  default     = "python-3.10"
```
same

andrewsykim commented:
> I prefer to keep this PR limited to changes relevant for the new notebook. Adding markdown to the older notebook wouldn't be too interesting since all the code is executed in one cell.

Should we simplify and remove the old notebook or keep both? With the changes in this PR, do both notebooks work?

richardsliu commented:
> > I prefer to keep this PR limited to changes relevant for the new notebook. Adding markdown to the older notebook wouldn't be too interesting since all the code is executed in one cell.
> 
> Should we simplify and remove the old notebook or keep both? With the changes in this PR, do both notebooks work?

I fixed some minor issues in the old notebook and added a header. We should probably keep it as a backup.

richardsliu commented:
/gcbrun

imreddy13 reviewed (approved):
Thank you!

imreddy13 commented:
/gcbrun
